### PR TITLE
feat(test): make simulation test deterministic!

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Deterministic simulation testing is available now! (on macOS only)

If you run the e2e simulation test with a specified seed:

```sh
MADSIM_TEST_SEED=1 ./risedev stest --release -p risingwave_simulation --no-capture
```

You'll find that it always outputs the exact same result:

```
...
[143.454969434s  INFO client sqllogictest::runner] test: Control(EndInclude("../../../e2e_test/streaming_delta_join/../tpch/drop_tables.slt.part"))
[143.454969578s ERROR frontend pgwire::pg_protocol] Error: connection closed
[143.454969578s ERROR frontend pgwire::pg_protocol] flush error: connection closed
[143.454969578s  INFO frontend pgwire::pg_server] Connection 192.168.100.1:1 closed
test basic ... ok
```

To make that possible, we intercepted several libc functions (such as `getrandom`, `clock_gettime`) to make them return predictable values. Application codes depending on `rand`, `std::time::{SystemTime, Instant}`, `std::collections::{HashMap, HashSet}` can benefit from it **without modification**! However, there is one exception: `rand` uses `syscall(SYS_getrandom, ...)` on Linux which is hard to hack. So deterministic testing is not available on Linux for now. (Fortunately, most of our developers are Mac users)

Note that this PR also upgrades the tonic version to v0.8. Now **it requires a system `protoc` as a dependency.**

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/source/src/manager.rs [hunk 1]
```diff
@@ -215,7 +215,7 @@ impl SourceManager for MemSourceManager {
             columns: source_columns,
             format: SourceFormat::Invalid,
             row_id_index: 0, // always use the first column as row_id
-            row_id_generator: Arc::new(Mutex::new(RowIdGenerator::with_epoch(
```
Comment (BugenZhao) [inline src/source/src/manager.rs:218]: So `std::sync::Mutex` does not support deterministic tests? 👀

Comment (wangrunji0408) [reply src/source/src/manager.rs:218]: `std::sync::MutexGuard` can not cross any `.await` point, or it will make the future `!Send`. 😇

Comment (BugenZhao) [reply src/source/src/manager.rs:218]: Where do we require `Send` before?

File: src/frontend/test_runner/tests/plan_test_runner.rs [hunk 2]
```diff
@@ -74,5 +74,9 @@ fn main() {
 
 #[cfg(madsim)]
 fn main() {
-    panic!("planner test is not supported yet in simulation");
+    // println!("planner test is not supported yet in simulation");
```
Comment (TennyZhuang) [inline src/frontend/test_runner/tests/plan_test_runner.rs:77]: Remove this?

**Comment by TennyZhuang** [comment]
Seems conflict with #4362

**Comment by wangrunji0408** [comment]
> Seems conflict with #4362

Let it get merged first 🥵

**Review by skyzh — APPROVED** [review]

**Review by TennyZhuang — APPROVED** [review]

File: src/tests/simulation/tests/e2e.rs [hunk 3]
```diff
@@ -60,23 +60,26 @@ async fn basic() {
...
-                "0.0.0.0:5688",
-                "--client-address",
-                "192.168.3.1:5688",
-                "--meta-address",
-                "192.168.1.1:5690",
-            ]);
-            risingwave_compute::start(opts).await
-        })
-        .build();
+    // TODO: support multiple nodes
```
Comment (BugenZhao) [inline src/tests/simulation/tests/e2e.rs:63]: May #4195 help this? 🥰

**Review by BugenZhao — APPROVED** [review]

File: src/tests/simulation/tests/e2e.rs (hunk 3 above)
Comment (wangrunji0408) [reply src/tests/simulation/tests/e2e.rs:63]: I've found it. Next time for sure! 🥰

**Pull request merged**